lib/resumption/client.go

```diff
@@ -0,0 +1,357 @@
+	}
+
+	hostID = string(match[2])
+
+	// discard is guaranteed to work for the line we just peeked
+	_, _ = conn.Discard(len(line))
+
+	return
+}
+
+type redialFunc = func(connCtx context.Context, hostID string) (net.Conn, error)
+
+// MaybeResumableSSHClientConn tries to detect if the server at the other end of
+// nc is a resumption v1 server, and if so it returns a [net.Conn] that will
+// transparently resume itself (using the provided redial func).
+func MaybeResumableSSHClientConn(nc net.Conn, connCtx context.Context, redial redialFunc) (net.Conn, error) {
+	dhKey, err := ecdh.P256().GenerateKey(rand.Reader)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to generate ECDH key, proceeding without resumption (this is a bug).")
+		return nc, nil
+	}
+
+	// adds a read buffer with the ability to peek to nc
+	conn := ensureMultiplexerConn(nc)
+
+	// We must send the first 8 bytes of the version string to go through some
+	// older versions of the multiplexer that sits in front of the Teleport SSH
+	// server; thankfully, no matter which SSH client we'll end up using, it
+	// must send `SSH-2.0-` as its first 8 bytes, as per RFC 4253 ("The Secure
+	// Shell (SSH) Transport Layer Protocol") section 4.2. Sending only the
+	// first 8 bytes rather than a full version string is noncompliant behavior,
+	// but our SSH client is only intended to talk to Teleport-implemented SSH
+	// servers anyway, and other clients in the ecosystem do much worse
+	// (ssh-keyscan will wait for the server to send data before sending
+	// anything, for example).
+	if _, err := conn.Write([]byte(sshPrefix)); err != nil {
+		conn.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	dhPub, hostID, err := readServerVersionExchange(conn)
+	if err != nil {
+		conn.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	if dhPub == nil {
+		// regular SSH connection, conn is about to read the SSH- line from the
+		// server but we've sent sshPrefix already, so we have to skip it from
+		// the application side writes
+		logrus.Debug("Server does not support resumption, proceeding without.")
+		return &sshVersionSkipConn{
+			Conn:           conn,
+			alreadyWritten: sshPrefix,
+		}, nil
+	}
+
+	dhSecret, err := dhKey.ECDH(dhPub)
+	if err != nil {
+		logrus.WithError(err).Warn("Failed to complete ECDH key exchange, proceeding without resumption.")
+		return &sshVersionSkipConn{
+			Conn:           conn,
+			alreadyWritten: sshPrefix,
+		}, nil
+	}
+
+	otp32 := sha256.Sum256(dhSecret)
+	token := resumptionToken(otp32[:16])
+
+	if _, err := conn.Write([]byte(clientSuffixV1)); err != nil {
+		conn.Close()
+		return nil, trace.Wrap(err)
+	}
+	if _, err := conn.Write(dhKey.PublicKey().Bytes()); err != nil {
+		conn.Close()
+		return nil, trace.Wrap(err)
+	}
+	if _, err := conn.Write([]byte{newConnClientExchangeTag}); err != nil {
+		conn.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	resumableConn := newResumableConn(conn.LocalAddr(), conn.RemoteAddr())
+
+	resumableConn.mu.Lock()
```

Comment:
Added a comment; the reason is that `runClientResumable` expects a brand new _locked_ `*Conn`, since nothing must interfere until the "first conn" attach is completed, and it's just easier to always require connections passed to `runResumeV1Unlocking` to be locked.